### PR TITLE
(ECA) TDNN + ivector support

### PR DESCRIPTION
**Pull request by noisychannel**
1. Modified the WSJ tdnn recipe to work with ECA
2. Used ivector based recipes from WSJ
3. Results included

PS : The results are not as great as a I expected (compared to ensemble training), especially on datasets that are slightly different (h5, sup). Ensemble training does well in comparison on h5 and sup.

### CODE REVIEW

File: egs/callhome_egyptian/s5/RESULTS [hunk 1]
```diff
@@ -166,3 +166,61 @@ exp/tri6a_dnn/decode_h5/wer_15 %WER 59.32 [ 9957 / 16785, 284 ins, 3960 del, 571
...
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_13 %WER 80.79 [ 14065 / 17409, 291 ins, 9155 del, 4619 sub ]
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_2 %WER 85.93 [ 14960 / 17409, 1330 ins, 6454 del, 7176 sub ]
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_3 %WER 84.65 [ 14737 / 17409, 1151 ins, 6635 del, 6951 sub ]
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_4 %WER 83.31 [ 14504 / 17409, 968 ins, 6890 del, 6646 sub ]
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_5 %WER 82.52 [ 14366 / 17409, 839 ins, 7159 del, 6368 sub ]
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_6 %WER 81.66 [ 14216 / 17409, 711 ins, 7477 del, 6028 sub ]
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_7 %WER 81.08 [ 14116 / 17409, 631 ins, 7750 del, 5735 sub ]
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_8 %WER 80.52 [ 14017 / 17409, 547 ins, 7999 del, 5471 sub ]
+exp/nnet3/nnet_tdnn_a/decode_sup/wer_9 %WER 80.31 [ 13982 / 17409, 468 ins, 8269 del, 5245 sub ]
+@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
```
Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: compared to your tri6a system these WERs are extremely high. Are you sure the TDNN+iVector system is working ??

Comment (danpovey) [inline egs/callhome_egyptian/s5/RESULTS]: Vijay, why don't you have a look at the diagnostics where he trained it?

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: @noisychannel  could you provide the location.

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: Here's the location : export/a04/gkumar/code/nnet3/egs/callhome_egyptian/s5

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: Also, I did not see anything unusual in the logs. It is strange that the system performs almost close to the ensemble system on dev/test but has drastically worse results on h5/sup.

Comment (danpovey) [inline egs/callhome_egyptian/s5/RESULTS]: Could be an instance where the iVectors fail due to domain mismach?  Maybe
play with silence-weight?  or try an iVector-free version of the system?
Dan

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: Training an ivector free system now.

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: for ivector free version you would remove <code> --online-ivector-dir </code> option use 

<code>  --cmvn-opts "--norm-means=true --norm-vars=true" </code>

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: Done. @vijayaditya did you see anything unusual with the ivector setup ?

Comment (danpovey) [inline egs/callhome_egyptian/s5/RESULTS]: You are getting way more deletions in that system, and fewer insertions,
compared with the normal system.
Vijay, is that anything like what we saw when we had weird iVectors in the
ASPIRE setup?
Dan


b06:s5: grep WER exp/nnet3/nnet_tdnn_a/decode_sup/wer_10
%WER 80.31 [ 13981 / 17409, 407 ins, 8528 del, 5046 sub ]
b06:s5: grep WER exp/tri nnet3/nnet_tdnn_a/decode_sup/wer_10
tri1/      tri1_ali/  tri2/      tri2_ali/  tri3a/     tri3a_ali/ tri4a/
  tri4a_ali/ tri5a/     tri5a_ali/ tri6a_dnn/
b06:s5: grep WER exp/tri5a/decode_sup/wer_10
%WER 63.77 [ 11101 / 17409, 1164 ins, 2476 del, 7461 sub ]

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: definitely a case of ivector mismatch. The average objective-function improvement in train and dev cases is ~5. While that _sup is varying between 24 and 120.

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: From the ASpIRE discussions:
We saw similar behavior in Aspire, you suggested using volume normalization which solved the problem to a great extent.

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: @noisychannel  to do volume normalization you would estimate the mean RMS value of training date using sox -n stat command. You would then scale the test data using sox --vol option. 

For an example look at the line 
https://github.com/kaldi-asr/kaldi/blob/master/egs/aspire/s5/local/multi_condition/aspire_data_prep.sh#L147

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: In h5, for example, the average obj-fun improvement varies drastically over different sets of frames : 
Average objective-function improvement was 3.66339 per frame, over 30901 frames (weighted).
vs
Average objective-function improvement was 144.811 per frame, over 29832 frames (weighted).

This is for the same dataset!

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: @noisychannel  measure the rms values of the files in training and test sets.

Comment (danpovey) [inline egs/callhome_egyptian/s5/RESULTS]: We also perturbed the training volumes.  Might it be a good idea,
regardless of this specific issue, to apply the script where we perturb the
training-data volumes and speeds?
Dan

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: volume and speed perturbation can be copied from
https://github.com/kaldi-asr/kaldi/blob/master/egs/swbd/s5c/local/nnet3/run_ivector_common.sh#L24

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: The average RMS values of the datasets:

train : .00391083750000000000
dev : .02442575000000000000
test : .01920605000000000000
sup : .01958655000000000000
h5 : .02346431578947368421

Obtained using:
```shell
while read line; do eval "$line sox -t wav - -n stat 2>&1 | grep 'RMS     amplitude:'"; done < <(head wav.scp | cut -d " " -f 2-) | awk '{print $3}' | paste -sd+ | bc | xargs -i echo {}"/"`wc -l wav.scp| awk '{print $1}'` | bc -l
```

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: Okay, both dev and test are fairly far off from train in their average RMS values. Yet, they seem to have decent results with ivectors. In contrast, h5 and sup have much worse performance with the same deviation from the train RMS. Thoughts?

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: You could decode with the current models, after normalizing the test data to be in the training data range.

Comment (danpovey) [inline egs/callhome_egyptian/s5/RESULTS]: That should help.  What would be even better is to train with training data
with perturbed volumes (hopefully including the range of the actual test
data).
The online-nnet2 system is supposed to learn the correct normalization from
the iVector, but it can't do so if there was no variation in that dimension
in the training data, i.e. if the training data was all normalized to the
same volume.
Dan

Comment (vijayaditya) [inline egs/callhome_egyptian/s5/RESULTS]: Yes, that would be the next step. Renormalizing the test data is just to avoid retraining the models, while verifying that normalization was the critical issue here.

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: Doing this now. 
1. Perturb the training data (volume + speed)
2. Extract MFCCs
3. Train ivector extractor and extract ivectors for all datasets
4. Train TDNN model with the train-ivectors and no CMVN
5. Decode test sets using the ivectors extracted in step 3.

Any additions?

PS : I'm updating the RESULTS now for training without ivectors. Things look much better (sup and h5 in the 50-55% WER range). However, this is still worse than ensemble training. Hopefully, perturbation can help.

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: Updated results : tri6a=Ensemble, nnet_tdnn_b=TDNN without ivectors

```
exp/tri6a_dnn/decode_dev/wer_9 %WER 50.31 [ 16859 / 33509, 1617 ins, 3794 del, 11448 sub ]
exp/tri6a_dnn/decode_test/wer_9 %WER 49.54 [ 7747 / 15637, 632 ins, 1873 del, 5242 sub ]
exp/tri6a_dnn/decode_sup/wer_9 %WER 58.15 [ 10123 / 17409, 1023 ins, 2477 del, 6623 sub ]
exp/tri6a_dnn/decode_h5/wer_9 %WER 55.76 [ 9360 / 16785, 666 ins, 2531 del, 6163 sub ]
```

```
exp/nnet3/nnet_tdnn_b/decode_dev/wer_9 %WER 49.86 [ 16706 / 33509, 1626 ins, 3866 del, 11214 sub ]
exp/nnet3/nnet_tdnn_b/decode_test/wer_9 %WER 49.94 [ 7809 / 15637, 595 ins, 1884 del, 5330 sub ]
exp/nnet3/nnet_tdnn_b/decode_sup/wer_9 %WER 59.66 [ 10387 / 17409, 1140 ins, 2415 del, 6832 sub ]
exp/nnet3/nnet_tdnn_b/decode_h5/wer_9 %WER 57.33 [ 9623 / 16785, 816 ins, 2558 del, 6249 sub ]
```

Comment (noisychannel) [inline egs/callhome_egyptian/s5/RESULTS]: Summary of a fairly long conversation with Vijay:
1. The ECA training data is small : ~14 hours
2. The current TDNN model has about 80% more parameters compared to the ensemble, but we could not find any indication of overfitting (4923458 vs 2933000)
3. Per frame accuracies are low for the TDNN model compared to the ensemble model (Really, they are both fairly low 25%-35%)
4. The train and validation accuracies seemed close to converging. More epochs may help though. 

Here's what I'm doing (in order):
1. Change TDNN splice indexes to "-2,-1,0,1,2 -1,2 -3,3 -7,2" (instead of the WSJ "-4,-3,-2,-1,0,1,2,3,4 0 -2,2 0 -4,4 0"). 4 layers also makes this more comparable to the ensemble model. Train and decode without ivectors : this should be a fair comparison against the ensemble model.
2. Train on speed and volume perturbed train set
3. Decode with and without RMS normalization for all models (this will benefit both the ensemble and the tDNN model)
4. (Future) : Use "800 relu-dim" instead to pnorm-input-dim and pnorm-output-dim

Thoughts/suggestions?

**Comment by danpovey** [comment]
@noisychannel , my assumption has been that this is not ready for merge and you are still making changes.  Let me know if I'm wrong.

**Comment by noisychannel** [comment]
@danpovey , as something that uses ensemble training and a baseline TDNN+ivector approach, this recipe is complete (with RESULTS included). If you wish, this can be merged with the master branch now. 

I'm trying out various configurations that may or may not be useful and can be included in the next commit (if they turn out to be useful).
